sqlglot/optimizer/canonicalize.py

```diff
@@ -70,17 +70,23 @@ def remove_redundant_casts(expression: exp.Expression) -> exp.Expression:
     return expression
 
 
+# this is designed for presto, there is a similar transform for tsql
+# this is different in that it only operates on int types, this is because
+# presto has a boolean type whereas tsql doesn't (people use bits)
+# with y as (select true as x) select x = 0 -- illegal presto query
 def ensure_bool_predicates(expression: exp.Expression) -> exp.Expression:
     if isinstance(expression, exp.Connector):
         _replace_int_predicate(expression.left)
         _replace_int_predicate(expression.right)
-
-    elif isinstance(expression, (exp.Where, exp.Having)) or (
+    elif isinstance(expression, exp.Not):
+        _replace_int_predicate(expression.this)
         # We can't replace num in CASE x WHEN num ..., because it's not the full predicate
-        isinstance(expression, exp.If)
-        and not (isinstance(expression.parent, exp.Case) and expression.parent.this)
+    elif isinstance(expression, exp.If) and not (
+        isinstance(expression.parent, exp.Case) and expression.parent.this
```

Comment:
it's not exactly the same, can the one in simplify be changed?

Reply:
i'll take a look

Reply:
ok i see what you mean, i guess let's leave the simplify condition as-is, but we can still dry this up so that it can be reused in the other section i commented in this pr